### PR TITLE
8327007: javax/swing/JSpinner/8008657/bug8008657.java fails

### PR DESCRIPTION
**Pull request by prsadhuk**
Test failed with the exception java.lang.IllegalArgumentException: (start <= value <= end) with no history of failing till date.
Investigation shows SpinnerDateModel constructor condition is not satisfied because
It seems it's a leap year problem that is being manifested..

As per test,
initDate Thu Feb 29 13:49:08 IST 2024 [ obtained by calendar.getTime();]
earliestDate Tue Feb 28 13:49:08 IST 2023 [obtained by calendar.add(Calendar.YEAR, -1);]
latestDate Wed Feb 28 13:49:08 IST 2024 [obtained by calendar.add(Calendar.YEAR, 1);]

Now, as per SpinnerDateModel constructor
earliestDate <= initDate condition is satisfied
but
latestDate >= initDate is not
so [start <= value <= end](https://github.com/openjdk/jdk/blob/998d0baab0fd051c38d9fd6021628eb863b80554/src/java.desktop/share/classes/javax/swing/SpinnerDateModel.java#L185) condition fails

Not sure it anything can be done in Calendar class for this but fix is done in test taking leapyear into account so that we add a day if it's a leapyear when we consecutively do Calendar.add(YEAR, -1) and Calendar.add(YEAR,1) on leapyear date of 29Feb

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8327007](https://bugs.openjdk.org/browse/JDK-8327007): javax/swing/JSpinner/8008657/bug8008657.java fails (**Bug** - P3)

### Reviewers
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18054/head:pull/18054` \
`$ git checkout pull/18054`

Update a local copy of the PR: \
`$ git checkout pull/18054` \
`$ git pull https://git.openjdk.org/jdk.git pull/18054/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18054`

View PR using the GUI difftool: \
`$ git pr show -t 18054`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18054.diff">https://git.openjdk.org/jdk/pull/18054.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18054#issuecomment-1970758013)

### CODE REVIEW

File: test/jdk/javax/swing/JSpinner/8008657/bug8008657.java [hunk 1]
```diff
@@ -134,9 +134,20 @@ static String getOrientation(ComponentOrientation orientation) {
     static void createDateSpinner() {
         Calendar calendar = Calendar.getInstance();
         Date initDate = calendar.getTime();
+        int year = calendar.get(Calendar.YEAR);
+        boolean isLeapYear = ((year % 4 == 0) && (year % 100 != 0) || (year % 400 == 0));
```
Comment (shurymury) [inline test/jdk/javax/swing/JSpinner/8008657/bug8008657.java:138]: Would it be better to use java.tim.Year.isLeap(long)?

**Comment by shurymury** [comment]
Can this test be changed to test multiple different dates, some of which are Feb 29th?

**Comment by prrace** [comment]
The problem comes down to this
% cat LeapDayBug.java 
import java.util.Calendar;

public class LeapDayBug {
   public static void main(String[] args) {
        Calendar calendar = Calendar.getInstance();
        calendar.set(2024, 1, 29);
        System.out.println(calendar.getTime());
        calendar.add(Calendar.YEAR, -1);
        calendar.add(Calendar.YEAR, 1);
        System.out.println(calendar.getTime());
   }
}

% java LeapDayBug.java 
Thu Feb 29 12:27:46 PST 2024
Wed Feb 28 12:27:46 PST 2024

So I'm not sure what the Calendar class could do about this since both
the -1 and the + 1 are correct considered standalone, but it just is not round trippable.

**Comment by prrace** [comment]
Surely this just needs a one line fix
diff --git a/test/jdk/javax/swing/JSpinner/8008657/bug8008657.java b/test/jdk/javax/swing/JSpinner/8008657/bug8008657.java
index cbeb0c185bb..49918278e34 100644
--- a/test/jdk/javax/swing/JSpinner/8008657/bug8008657.java
+++ b/test/jdk/javax/swing/JSpinner/8008657/bug8008657.java
@@ -137,6 +137,7 @@ static void createDateSpinner() {
         calendar.add(Calendar.YEAR, -1);
         Date earliestDate = calendar.getTime();
         calendar.add(Calendar.YEAR, 1);
+        calendar.add(Calendar.DAY_OF_MONTH, 1);
         Date latestDate = calendar.getTime();
         SpinnerModel dateModel = new SpinnerDateModel(initDate,
                 earliestDate,

**Comment by prsadhuk** [comment]
> Surely this just needs a one line fix
> 
> diff --git a/test/jdk/javax/swing/JSpinner/8008657/bug8008657.java b/test/jdk/javax/swing/JSpinner/8008657/bug8008657.java
> index cbeb0c185bb..49918278e34 100644
> --- a/test/jdk/javax/swing/JSpinner/8008657/bug8008657.java
> +++ b/test/jdk/javax/swing/JSpinner/8008657/bug8008657.java
> @@ -137,6 +137,7 @@ static void createDateSpinner() {
>          calendar.add(Calendar.YEAR, -1);
>          Date earliestDate = calendar.getTime();
>          calendar.add(Calendar.YEAR, 1);
> +        calendar.add(Calendar.DAY_OF_MONTH, 1);
>          Date latestDate = calendar.getTime();
>          SpinnerModel dateModel = new SpinnerDateModel(initDate,
>                  earliestDate,

Yes, I did thought about that but thought should we do it irrespectively of leap year or not, so atlast I thought of doing it for leapyear only so that it doesn't affect anyother scenarios...
But if you want to add 1 day to it irrespectively and cutshort the added code, I can do it too...

**Comment by prsadhuk** [comment]
> Surely this just needs a one line fix
> 
> diff --git a/test/jdk/javax/swing/JSpinner/8008657/bug8008657.java b/test/jdk/javax/swing/JSpinner/8008657/bug8008657.java
> index cbeb0c185bb..49918278e34 100644
> --- a/test/jdk/javax/swing/JSpinner/8008657/bug8008657.java
> +++ b/test/jdk/javax/swing/JSpinner/8008657/bug8008657.java
> @@ -137,6 +137,7 @@ static void createDateSpinner() {
>          calendar.add(Calendar.YEAR, -1);
>          Date earliestDate = calendar.getTime();
>          calendar.add(Calendar.YEAR, 1);
> +        calendar.add(Calendar.DAY_OF_MONTH, 1);
>          Date latestDate = calendar.getTime();
>          SpinnerModel dateModel = new SpinnerDateModel(initDate,
>                  earliestDate,

Done minimal change as suggested...

**Review by prrace — APPROVED** [review]

**Pull request closed without merging**